### ISSUE TITLE
Bind Nicotine+ to specified network adapter

### ISSUE THREAD

**Issue by dochdoch** [nicotine-plus/nicotine-plus#871]
I'd like to be able to choose a network adapter to bind Nicotine+ to.

This is useful for VPNs, etc. qBittorrent has implemented this very nicely as an advanced preference via a simple dropdown menu.

**Comment by gits68** [comment]
much better would be to have the possibility to bind the incoming & outcoming interfaces to différent network adapter.
for example, I've an ADSL and a 4G network, can't use the 4G network for download bcoz it is not possible to listen on it...
while it would be possible to download using the 4G network and upload using the ADSL one...

**Comment by felipehw** [comment]
I agree with @dochdoch .
Actually, I need to use `nicotine  -b VpnIP` from CLI.  I always need to look at the actual IP from the VPN interface before invoking the command.
If we could (just as we do with qBittorrent) just set the interface from the GUI ... these boring procedures vanish.

**Comment by mathiascode** [comment]
I assume this option is to ensure that only the VPN is used for traffic at all times, or is there more to it? Are there cases where the VPN isn't used by default? I'm not too familiar with this area myself, so any insight or contributions would be appreciated.

**Comment by felipehw** [comment]
@mathiascode 

If your VPN turns off and you don't perceive ... Nicotine connects without VPN.
If we can attach Nicotine to an interface (like `tun0`), this kind of problem doesn't happen.

**Comment by mathiascode** [comment]
This is now done for Linux and macOS. *BSD and Windows will require more work.

**Comment by dochdoch** [comment]
Thank you very much for all your hard work!

**Comment by felipehw** [comment]
@mathiascode Thanks a lot man!

**Comment by mathiascode** [comment]
https://github.com/nicotine-plus/nicotine-plus/commit/5949bf0bd7bea04b8fd3b300e8b162a9a673ccd8 should add this functionality on *BSD. Would be much appreciated if any *BSD users could test this.

**Comment by slook** [comment]
To test binding to a particular network adapter with Preferences > Network > Network Interface in the [latest unstable build](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md) of master branch 3.3.0.dev4

**Comment by mathiascode** [comment]
Confirmed working on OpenBSD: https://github.com/nicotine-plus/nicotine-plus/commit/5949bf0bd7bea04b8fd3b300e8b162a9a673ccd8#commitcomment-92062623

I've also verified that it works on FreeBSD.

The only part remaining of this feature request is Windows support, more specifically a Windows replacement for this method (without external dependencies): https://github.com/nicotine-plus/nicotine-plus/blob/42bfd32aa4d2bd1b8fae52bdd7e100148c7d4c82/pynicotine/slskproto.py#L320-L336

**Comment by slook** [comment]
Can we use some sort of `netstat` stdout hack similar to that recently implemented in the portmapper.py?

**Comment by mathiascode** [comment]
IIRC, Windows has several APIs to achieve what we want. We use this Python function to receive a list of available network interfaces: https://docs.python.org/3/library/socket.html#socket.if_nameindex

What we need to do is receive the IP address of the interface from the interface name (e.g. `ethernet_32770`, not the UUID or friendly name).

**Comment by slook** [comment]
https://github.com/pydron/ifaddr/blob/master/ifaddr/_win32.py

**Comment by mathiascode** [comment]
Due to how the Win32 API works, I decided to not use `socket.if_nameindex`, and instead show the friendly interface names in the GUI. qBittorrent uses these too.

**Comment by slook** [comment]
@dochdoch @felipehw 
PR #2925 might cause a regression to the fix which was implemented prior in fix to this issue. Latest master 3.3.3dev1 needs to be tested in the wild to verify that fallback doesn't happen in the case of any sort of network outage or other error, link:

https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md

**Comment by jult** [comment]
I don't see this in windows n+ client, has it been removed again? I did try
p:\Portable\Nicotine+\Nicotine+.exe -b 192.168.1.100
but it still shows the public IP. I need to tunnel it through ssh remotely because of CGNAT.

**Comment by slook** [comment]
Windows was late to get the network adapter binding feature so it is only fully available in very recent versions, try 3.4.0.dev1...
https://nicotine-plus.org/doc/TESTING.html#windows

You can set the a particular interface in Preferences -> Network -> Connections -> "Network Interface", or you can set by IP address on the command line.

In the latest builds it's not necessary to restart the application for the network settings to take effect immediately on pressing Apply or OK in the Preferences dialog, whereas the initial implementation in the 3.3.x versions couldn't do that.

**Comment by jult** [comment]
Yeah, I tried those versions, but they fail:

<img width="385" height="155" alt="Image" src="https://github.com/user-attachments/assets/b98b6fb1-3da3-4275-ac2e-f07268774a41" />

<img width="412" height="1052" alt="Image" src="https://github.com/user-attachments/assets/53e7b302-e710-4b8f-829e-d69a707293c4" />

P:\Portable\Nicotine+\Nicotine+-debug.exe:

Windows fatal exception: code 0xc0000139

Thread 0x00003410 (most recent call first):
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/module.py", line 145 in __getattr__
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/overrides/Gdk.py", line 99 in <module>
  File "<frozen importlib._bootstrap>", line 491 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 763 in exec_module
  File "<frozen importlib._bootstrap>", line 938 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1342 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1371 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1406 in _gcd_import
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88 in import_module
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/overrides/__init__.py", line 128 in load_overrides
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 145 in create_module
  File "<frozen importlib._bootstrap>", line 816 in module_from_spec
  File "<frozen importlib._bootstrap>", line 924 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1342 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1371 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1406 in _gcd_import
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88 in import_module
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 144 in create_module
  File "<frozen importlib._bootstrap>", line 816 in module_from_spec
  File "<frozen importlib._bootstrap>", line 924 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1342 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1371 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1406 in _gcd_import
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88 in import_module
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 144 in create_module
  File "<frozen importlib._bootstrap>", line 816 in module_from_spec
  File "<frozen importlib._bootstrap>", line 924 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1342 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1371 in _find_and_load
  File "<frozen importlib._bootstrap>", line 491 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1434 in _handle_fromlist
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/gtkgui/__init__.py", line 84 in check_gtk_version
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/gtkgui/__init__.py", line 194 in run
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/__init__.py", line 218 in run
  File "D:/a/nicotine-plus/nicotine-plus/nicotine", line 20 in load_module
  File "D:/a/nicotine-plus/nicotine-plus/nicotine", line 30 in <module>
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/freeze_core/initscripts/console.py", line 43 in run
  File "src/freeze_core/initscripts/__startup__.py", line 147 in run

Current thread's C stack trace (most recent call first):
  <cannot get C stack on this system>

(process:22276): GLib-GIRepository-WARNING **: 21:52:30.729: Failed to load shared library 'libgtk-4-1.dll' referenced by the typelib: 'libgtk-4-1.dll': The specified procedure could not be found.
Traceback (most recent call last):
  File "src/freeze_core/initscripts/__startup__.py", line 147, in run
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/freeze_core/initscripts/console.py", line 43, in run
  File "D:/a/nicotine-plus/nicotine-plus/nicotine", line 30, in <module>
  File "D:/a/nicotine-plus/nicotine-plus/nicotine", line 20, in load_module
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/__init__.py", line 218, in run
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/gtkgui/__init__.py", line 194, in run
  File "D:/a/nicotine-plus/nicotine-plus/pynicotine/gtkgui/__init__.py", line 84, in check_gtk_version
  File "<frozen importlib._bootstrap>", line 1371, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1342, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 924, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 816, in module_from_spec
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 144, in create_module
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88, in import_module
  File "<frozen importlib._bootstrap>", line 1406, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1371, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1342, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 924, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 816, in module_from_spec
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 144, in create_module
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88, in import_module
  File "<frozen importlib._bootstrap>", line 1406, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1371, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1342, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 924, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 816, in module_from_spec
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/importer.py", line 145, in create_module
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/overrides/__init__.py", line 128, in load_overrides
  File "D:/M/msys64/clang64/lib/python3.14/importlib/__init__.py", line 88, in import_module
  File "D:/a/_temp/msys64/clang64/lib/python3.14/site-packages/gi/overrides/Gdk.py", line 460, in <module>
AttributeError: type object 'FileList' has no attribute 'STATIC_SIZE'

**Comment by slook** [comment]
Thanks please could you paste the plain text traceback and the screenshot of the smaller message box into your new issue.

**Comment by slook** [comment]
Since the crash has now been fixed thanks to your report, you should now be able to try the improved port forwarding again in the latest build at https://nicotine-plus.org/doc/TESTING.html#windows

**Comment by mathiascode** [comment]
I believe I added all the connection-related improvements to 3.3.x before we started development of 3.4.x, so using the latest stable 3.3.11 release should be enough.
